### PR TITLE
feat: support compressed csv outputs

### PR DESCRIPTION
**Pull request by bloodearnest**
Ideally, we'd only support compressed outputs, but that would be
a backwards incompatible change.

### CODE REVIEW

File: sqlrunner/main.py [hunk 1]
```diff
@@ -87,10 +89,18 @@ def write_results(results, f_path):
         return
 
     fieldnames = first_result.keys()
-    with f_path.open(mode="w", newline="", encoding="utf-8") as f:
+
+    if f_path.suffixes == [".csv", ".gz"]:
+        f = gzip.open(f_path, "wt", newline="", compresslevel=6)
```
Comment (madwort) [inline sqlrunner/main.py:94]: does this need explicit `encoding="utf-8"`, the same as the other branch?

**Review by madwort — APPROVED** [review]

File: sqlrunner/main.py (hunk 1 above)
Comment (bloodearnest) [reply sqlrunner/main.py:94]: The t in "wt" implies encoding in utf-8, but its probably better to be explicit

**Comment by iaindillingham** [comment]
Thank you 🙂

**Review by iaindillingham — APPROVED** [review]

File: sqlrunner/main.py (hunk 1 above)
Comment (madwort) [reply sqlrunner/main.py]: ah

**Comment by benbc** [comment]
How many studies do we have using SQL Runner? Would it really be bad to make it backwards compatible and change them? This is the kind of migration that is annoying to do but becomes enormously more expensive the longer you leave it.

**Comment by bloodearnest** [comment]
> How many studies do we have using SQL Runner? Would it really be bad to make it backwards compatible and change them? This is the kind of migration that is annoying to do but becomes enormously more expensive the longer you leave it.

I don't think that many, and they're all internal Bennett ones? Easier to do now than later, too.

It wouldn't be very much work to change their project.yaml and code to switch to csv.gz.  It would also need some clean up of workspaces to get the space saving benefit however, as otherwise the uncompressed csv files will live forever.

This is a fundamental quirk of zip compression changing the filename. `.feather`, `.arrow`, `.rds` don't have this problem, as optional compression is built in to the format, the name doesn't change.

Actually, if we're going to change it everywhere could we change it to feather?

**Comment by iaindillingham** [comment]
I think there are less than 10 studies, and they are all internal. They're not run that often, if at all, and don't output large files. The Winter Pressures study was the exception and prompted a previous change to SQL Runner, to support streaming (#61).

Updating these studies wouldn't be onerous, but I also don't think it's necessary in most cases. I will change the Winter Pressures study today.

I've found some quirks with Arrow that I'd like to investigate before supporting it.

**Comment by iaindillingham** [comment]
Another reason for not defaulting to compressed csv is that in most cases, again excepting Winter Pressures, we want to be able to eyeball the file on L4 through OS Jobs. We couldn't do that with compressed csv, without uncompressing it; and we couldn't do that with Arrow.[^1]

[^1]: That is, I don't think the viewer on OS Jobs supports either compressed csv or Arrow.

**Comment by benbc** [comment]
@iaindillingham do you think it's not necessary to update those studies because the files don't take much space or because it doesn't matter if they don't work with the latest version of SQL Runner?

**Comment by benbc** [comment]
> It wouldn't be very much work to change their project.yaml and code to switch to csv.gz. It would also need some clean up of workspaces to get the space saving benefit however, as otherwise the uncompressed csv files will live forever.

@bloodearnest I don't think that the space saving for these studies matters (except possibly the winter pressures one). The reason for making zipping mandatory is not to save space now but to make it impossible for people to do the wrong thing in the future.

**Comment by iaindillingham** [comment]
> @iaindillingham do you think it's not necessary to update those studies because the files don't take much space or because it doesn't matter if they don't work with the latest version of SQL Runner?

Because the files don't take much space; and because we want to eyeball the files on L4 through OS Jobs.

I think the definition of "the wrong thing" is different here. I also think using SQL Runner within a study, where there is the potential to write large files, should also be as exceptional as exceptional can be.

**Comment by bloodearnest** [comment]
> > It wouldn't be very much work to change their project.yaml and code to switch to csv.gz. It would also need some clean up of workspaces to get the space saving benefit however, as otherwise the uncompressed csv files will live forever.
> 
> @bloodearnest I don't think that the space saving for these studies matters (except possibly the winter pressures one). The reason for making zipping mandatory is not to save space now but to make it impossible for people to do the wrong thing in the future.

An current rate of disk consumption, we've got about 6 months of runway before we start hitting problems again. If there's only one sqlrunner study that's big, fair enough, but it all addes up.

**Comment by bloodearnest** [comment]
> I think the definition of "the wrong thing" is different here. I also think using SQL Runner within a study, where there is the potential to write large files, should also be as exceptional as exceptional can be.

That's a fair point. We don't expect sqlrunner to be used frequently, or for non-curation work, which we wouldn't expect to use lots of L3 data.

**Comment by bloodearnest** [comment]
> Another reason for not defaulting to compressed csv is that in most cases, again excepting Winter Pressures, we want to be able to eyeball the file on L4 through OS Jobs. We couldn't do that with compressed csv, without uncompressing it; and we couldn't do that with Arrow.[1](#user-content-fn-1-ca45fc4edcec5a4847570a8ae960b76d)

So, we've discussed implementing a file limit for files to be copied to L4, and there is already a size limit for what L4 files can be released.

But this is a good point - unlike cohortextractor, who's outputs are nearly always level 3, sqlrunner can do either level 3 or level 4 outputs. Indeed, it's main initial use case is for curation work, which is probably producing level 4 outputs. Given that, the above implementation allowing the user to choose is probably right?

> ## Footnotes
> 1. That is, I don't think the viewer on OS Jobs supports either compressed csv or Arrow. [leftwards_arrow_with_hook](#user-content-fnref-1-ca45fc4edcec5a4847570a8ae960b76d)

Not currently, but I hope we can fix that soon as part of the SACRO work.

**Pull request merged**